
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>VAJA: Programski dostop do GenBank (Python) &#8212; Biokemijska informatika</title>
    
  <link href="../_static/css/theme.css" rel="stylesheet" />
  <link href="../_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="canonical" href="https://mpavsic.github.io/biokeminfo/vaje/genbank_eutilities.html" />
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Aminokislinska zaporedja" href="../osnove/ak.html" />
    <link rel="prev" title="VAJA: Dostop do GenBank v brskalniku" href="genbank_web.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Biokemijska informatika</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../naslovnica.html">
   Uvod
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Zbirke
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../osnove/pubmed.html">
   PubMed
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="pubmed_web.html">
     VAJA: Iskanje po PubMed
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="pubmed_eutilities.html">
     DEMO: Programski dostop do PubMed
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../osnove/nt.html">
   Nukleotidna zaporedja
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="nt-oligo_analiza.html">
     VAJA: Enostavna analiza nt-zaporedja (Python)
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="genbank_web.html">
     VAJA: Dostop do GenBank v brskalniku
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     VAJA: Programski dostop do GenBank (Python)
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../osnove/ak.html">
   Aminokislinska zaporedja
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="uniprot_web.html">
     VAJA: Dostop do UniProt v brskalniku
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="uniprot_rest.html">
     VAJA: Programski dostop do UniProt (Python)
    </a>
   </li>
  </ul>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Poravnave
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../osnove/dotplot.html">
   Točkovni diagram
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="dotplot_web.html">
     VAJA: Točkovni diagram
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../osnove/poravnave.html">
   Poravnave
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="matrika_zamenjav.html">
     VAJA: Izračun matrike zamenjav (Python)
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="poravnava2_web.html">
     VAJA: Poravnava zaporedij NW &amp; SW (spletna orodja)
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="poravnava2_nw_sw.html">
     VAJA: Poravnava zaporedij NW &amp; SW (Python)
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="poravnava2_lastna_matrika.html">
     DEMO: Izračun lastne matrike in njena uporaba
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="blast.html">
     VAJA: Iskanje podobnih zaporedij z BLAST
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="msa.html">
     VAJA: Poravnave več zaporedij
    </a>
   </li>
  </ul>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Grafi
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../osnove/grafi.html">
   Grafi v molekularni biologiji
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="grafi_cytoscape.html">
     VAJA: Grafi in program Cytoscape
    </a>
   </li>
  </ul>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Strukture
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../osnove/strukture.html">
   Strukture
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="podobne_strukture.html">
     VAJA: Iskanje podobnih struktur
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="modeliranje.html">
     VAJA: Modeliranje na osnovi homologije
    </a>
   </li>
  </ul>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Priloge
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../priloge/programi.html">
   Programi
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../priloge/python.html">
     Python
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../priloge/jupyterlab.html">
     JupyterLab
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../priloge/markdown.html">
     Markdown
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../priloge/git.html">
     Git in GitHub
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../priloge/datoteke.html">
   Poimenovanje datotek
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../priloge/drobnarije.html">
   Drobnarije
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Izvedba
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../izvedba/vaje_2020-2021.html">
   Vaje 2020/2021
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/vaje/genbank_eutilities.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/mpavsic/biokeminfo"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/mpavsic/biokeminfo/main?urlpath=lab/tree/biokeminfo/vaje/genbank_eutilities.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        <a class="colab-button" href="https://colab.research.google.com/github/mpavsic/biokeminfo/blob/main/biokeminfo/vaje/genbank_eutilities.ipynb"><button type="button" class="btn btn-secondary topbarbtn"
                title="Launch Colab" data-toggle="tooltip" data-placement="left"><img class="colab-button-logo"
                    src="../_static/images/logo_colab.png"
                    alt="Interact on Colab">Colab</button></a>
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#zgledi">
   Zgledi
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#dostop-do-genbank">
     Dostop do GenBank
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#shranjevanje-zapisa-v-datoteko">
     Shranjevanje zapisa v datoteko
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ekstrakcija-dela-zapisa-in-prevajanje">
     Ekstrakcija dela zapisa in prevajanje
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#naloga">
   Naloga
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ozadje">
     Ozadje
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     Naloga
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#namigi">
     Namigi
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#v-razmislek">
     V razmislek
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="vaja-programski-dostop-do-genbank-python">
<h1>VAJA: Programski dostop do GenBank (Python)<a class="headerlink" href="#vaja-programski-dostop-do-genbank-python" title="Permalink to this headline">¶</a></h1>
<p>Tudi pri tej vaji bomo, podobno kot pri dostopu do PubMeda, uporabljali <a class="reference external" href="https://biopython.org">BioPython</a>. Navodila z zgledi so na voljo v <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html">BioPython Tutorial and Cookbook</a>.</p>
<p>Najprej si oglejte in preizkusite delovanje delov kode v zgledih, nato pa z ustrezno kombinacijo nekaterih delov kode iz zgledov ter prilagoditvijo kode rešite nalogo na dnu zvezka.</p>
<hr class="docutils" />
<div class="section" id="zgledi">
<h2>Zgledi<a class="headerlink" href="#zgledi" title="Permalink to this headline">¶</a></h2>
<div class="section" id="dostop-do-genbank">
<h3>Dostop do GenBank<a class="headerlink" href="#dostop-do-genbank" title="Permalink to this headline">¶</a></h3>
<p>Dostop do GenBank je mogoč direktno z uporabo modulov v <strong>BioPython</strong>, podobno kot to velja za PubMed (modul Bio.Entrez, ki smo ga spoznali <a class="reference internal" href="pubmed_eutilities.html"><span class="doc std std-doc">tukaj</span></a>). Tudi tokrat se moramo za uporabo E-utilities identificirati z e-poštnim naslovom. Če <em>accession code</em> poznamo, lahko uporabimo funkcijo <code class="docutils literal notranslate"><span class="pre">efetch</span></code>, sicer pa izvedemo iskanje z <code class="docutils literal notranslate"><span class="pre">esearch</span></code> in šele nato prenesemo zapis(e) z <code class="docutils literal notranslate"><span class="pre">efetch</span></code>, v obeh primerih pa rezultate preberemo z <code class="docutils literal notranslate"><span class="pre">read</span></code>.</p>
<p>Pa kar gremo k primeru, ki ilustrira branje enega zapisa - tako torej dostopamo do zapisa, če poznamo <strong>kodo za dostop</strong> (<em>accession code</em>). Če ne napišemo verzije zapisa nam vrne zadnjo (najnovejšo) verzijo, sicer pa lahko verzijo sami določimo tako, da <em>accession code</em> napišemo v obliki NM_005985.4 (če je osnovna koda NM_005985).</p>
<p>Kot zbirko (<code class="docutils literal notranslate"><span class="pre">db</span></code>) izberemo <em>nucleotide</em>, podatke pa prenesemo v obliki teksta (<code class="docutils literal notranslate"><span class="pre">retmod='text'</span></code>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>

<span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s1">&#39;sem.napisite.vas@e-postni.naslov&#39;</span> <span class="c1"># identificiramo se</span>

<span class="n">acode</span> <span class="o">=</span> <span class="s1">&#39;NM_005985&#39;</span>   <span class="c1"># koda zapisa</span>

<span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s1">&#39;nucleotide&#39;</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s1">&#39;fasta&#39;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s1">&#39;text&#39;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">acode</span><span class="p">)</span>
<span class="n">seq_record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> z </span><span class="si">%i</span><span class="s1"> značilnostmi&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">features</span><span class="p">)))</span>   <span class="c1"># koliko značilnosti je?</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>   <span class="c1"># izpis zaporedja</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seq_record</span><span class="p">)</span>   <span class="c1"># izpis celotnega objekta</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>NM_005985.4 z 0 značilnostmi
ATTGCGCCGCGGCACGGCCTAGCGAGTGGTTCTTCTGCGCTACTGCTGCGCGAATCGGCGACCCCAGTGCCTCGACCACTATGCCGCGCTCTTTCCTCGTCAGGAAGCCCTCCGACCCCAATCGGAAGCCTAACTACAGCGAGCTGCAGGACTCTAATCCAGAGTTTACCTTCCAGCAGCCCTACGACCAGGCCCACCTGCTGGCAGCCATCCCACCTCCGGAGATCCTCAACCCCACCGCCTCGCTGCCAATGCTCATCTGGGACTCTGTCCTGGCGCCCCAAGCCCAGCCAATTGCCTGGGCCTCCCTTCGGCTCCAGGAGAGTCCCAGGGTGGCAGAGCTGACCTCCCTGTCAGATGAGGACAGTGGGAAAGGCTCCCAGCCCCCCAGCCCACCCTCACCGGCTCCTTCGTCCTTCTCCTCTACTTCAGTCTCTTCCTTGGAGGCCGAGGCCTATGCTGCCTTCCCAGGCTTGGGCCAAGTGCCCAAGCAGCTGGCCCAGCTCTCTGAGGCCAAGGATCTCCAGGCTCGAAAGGCCTTCAACTGCAAATACTGCAACAAGGAATACCTCAGCCTGGGTGCCCTCAAGATGCACATCCGAAGCCACACGCTGCCCTGCGTCTGCGGAACCTGCGGGAAGGCCTTCTCTAGGCCCTGGCTGCTACAAGGCCATGTCCGGACCCACACTGGCGAGAAGCCCTTCTCCTGTCCCCACTGCAGCCGTGCCTTCGCTGACCGCTCCAACCTGCGGGCCCACCTCCAGACCCACTCAGATGTCAAGAAGTACCAGTGCCAGGCGTGTGCTCGGACCTTCTCCCGAATGTCCCTGCTCCACAAGCACCAAGAGTCCGGCTGCTCAGGATGTCCCCGCTGACCCTCGAGGCTCCCTCTTCCTCTCCATACCTGCCCCTGCCTGACAGCCTTCCCCAGCTCCAGCAGGAAGGACCCCACATCCTTCTCACTGCCATGGAATTCCCTCCTGAGTGCCCCACTTCTGGCCACATCAGCCCCACAGGACTTTGATGAAGACCATTTTCTGGTTCTGTGTCCTCTGCCTGGGCTCTGGAAGAGGCCTTCCCATGGCCATTTCTGTGGAGGGAGGGCAGCTGGCCCCCAGCCCTGGGGGATTCCTGAGCTGGCCTGTCTGCGTGGGTTTTTGTATCCAGAGCTGTTTGGATACAGCTGCTTTGAGCTACAGGACAAAGGCTGACAGACTCACTGGGAAGCTCCCACCCCACTCAGGGGACCCCACTCCCCTCACACACACCCCCCCACAAGGAACCCTCAGGCCACCCTCCACGAGGTGTGACTAACTATGCAATAATCCACCCCCAGGTGCAGCCCCAGGGCCTGCGGAGGCGGTGGCAGACTAGAGTCTGAGATGCCCCGAGCCCAGGCAGCTATTTCAGCCTCCTGTTTGGTGGGGTGGCACCTGTTTCCCGGGCAATTTAACAATGTCTGAAAAGGGACTGTGAGTAATGGCTGTCACTTGTCGGGGGCCCAAGTGGGGTGCTCTGGTCTGACCGATGTGTCTCCCAGAACTATTCTGGGGGCCCGACAGGTGGGCCTGGGAGGAAGATGTTTACATTTTTAAAGGTACACTGGTATTTATATTTCAAACATTTTGTATCAAGGAAACGTTTTGTATAGTTATATGTACAGTTTATTGATATTCAATAAAGCAGTTAATTTATATATTAAAAA
ID: NM_005985.4
Name: NM_005985.4
Description: NM_005985.4 Homo sapiens snail family transcriptional repressor 1 (SNAI1), mRNA
Number of features: 0
Seq(&#39;ATTGCGCCGCGGCACGGCCTAGCGAGTGGTTCTTCTGCGCTACTGCTGCGCGAA...AAA&#39;)
</pre></div>
</div>
</div>
</div>
<p>Vidimo, da smo pridobili zapis, a nima nič značilnosti, spodaj pa je zaporedje. Zakaj ni nič značilnosti, če pa jih najdemo na GenBank? Povezava do zapisa: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/nuccore/NM_005985">https://www.ncbi.nlm.nih.gov/nuccore/NM_005985</a> (značilnosti so v delu <em>features</em>).</p>
<p>Mi značilnosti nismo pridobili zato, ker smo zapis prenesli v formatu FASTA. Če zadevo ponovimo v formatu GenBank (gb)…</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s1">&#39;nucleotide&#39;</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s1">&#39;gb&#39;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s1">&#39;text&#39;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">acode</span><span class="p">)</span>
<span class="n">seq_record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s1">&#39;gb&#39;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> z </span><span class="si">%i</span><span class="s1"> značilnostmi&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">features</span><span class="p">)))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seq_record</span><span class="p">)</span>   <span class="c1"># izpis celotnega objekta</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>NM_005985.4 z 27 značilnostmi
ID: NM_005985.4
Name: NM_005985
Description: Homo sapiens snail family transcriptional repressor 1 (SNAI1), mRNA
Number of features: 27
/molecule_type=mRNA
/topology=linear
/data_file_division=PRI
/date=21-FEB-2021
/accessions=[&#39;NM_005985&#39;]
/sequence_version=4
/keywords=[&#39;RefSeq&#39;, &#39;MANE Select&#39;]
/source=Homo sapiens (human)
/organism=Homo sapiens
/taxonomy=[&#39;Eukaryota&#39;, &#39;Metazoa&#39;, &#39;Chordata&#39;, &#39;Craniata&#39;, &#39;Vertebrata&#39;, &#39;Euteleostomi&#39;, &#39;Mammalia&#39;, &#39;Eutheria&#39;, &#39;Euarchontoglires&#39;, &#39;Primates&#39;, &#39;Haplorrhini&#39;, &#39;Catarrhini&#39;, &#39;Hominidae&#39;, &#39;Homo&#39;]
/references=[Reference(title=&#39;The expression and clinical significance of an Epithelial-Mesenchymal Transition inducer, SNAI1, in head and neck carcinoma&#39;, ...), Reference(title=&#39;Deubiquitinase USP29 promotes gastric cancer cell migration by cooperating with phosphatase SCP1 to stabilize Snail protein&#39;, ...), Reference(title=&#39;Snail promotes prostate cancer migration by facilitating SPOP ubiquitination and degradation&#39;, ...), Reference(title=&#39;The New Model of Snail Expression Regulation: The Role of MRTFs in Fast and Slow Endothelial-Mesenchymal Transition&#39;, ...), Reference(title=&#39;Wnt/CTNNB1 Signal Transduction Pathway Inhibits the Expression of ZFP36 in Squamous Cell Carcinoma, by Inducing Transcriptional Repressors SNAI1, SLUG and TWIST&#39;, ...), Reference(title=&#39;Down-regulation of promoter 1.3 activity of the human aromatase gene in breast tissue by zinc-finger protein, snail (SnaH)&#39;, ...), Reference(title=&#39;Identification and analysis of two snail genes in the pufferfish (Fugu rubripes) and mapping of human SNA to 20q&#39;, ...), Reference(title=&#39;The transcription factor snail is a repressor of E-cadherin gene expression in epithelial tumour cells&#39;, ...), Reference(title=&#39;Genomic organization, expression, and chromosome location of the human SNAIL gene (SNAI1) and a related processed pseudogene (SNAI1P)&#39;, ...), Reference(title=&#39;Characterisation of the human snail (SNAI1) gene and exclusion as a major disease gene in craniosynostosis&#39;, ...)]
/comment=REVIEWED REFSEQ: This record has been curated by NCBI staff. The
reference sequence was derived from DA972913.1 and BC012910.1.
On Nov 22, 2018 this sequence version replaced NM_005985.3.
Summary: The Drosophila embryonic protein snail is a zinc finger
transcriptional repressor which downregulates the expression of
ectodermal genes within the mesoderm. The nuclear protein encoded
by this gene is structurally similar to the Drosophila snail
protein, and is also thought to be critical for mesoderm formation
in the developing embryo. At least two variants of a similar
processed pseudogene have been found on chromosome 2. [provided by
RefSeq, Jul 2008].
Publication Note:  This RefSeq record includes a subset of the
publications that are available for this gene. Please see the Gene
record to access additional publications.
COMPLETENESS: full length.
/structured_comment=OrderedDict([(&#39;Evidence-Data&#39;, OrderedDict([(&#39;Transcript exon combination&#39;, &#39;BC012910.1, SRR5189661.136537.1 [ECO:0000332]&#39;), (&#39;RNAseq introns&#39;, &#39;single sample supports all introns SAMEA1965299, SAMEA1966682 [ECO:0000348]&#39;)])), (&#39;RefSeq-Attributes&#39;, OrderedDict([(&#39;MANE Ensembl match&#39;, &#39;ENST00000244050.3/ ENSP00000244050.2&#39;), (&#39;RefSeq Select criteria&#39;, &#39;based on single protein-coding transcript&#39;)]))])
Seq(&#39;ATTGCGCCGCGGCACGGCCTAGCGAGTGGTTCTTCTGCGCTACTGCTGCGCGAA...AAA&#39;)
</pre></div>
</div>
</div>
</div>
<p>Če preštejemo značilnosti v zapisu na spletu vidimo, da jih je dejansko toliko. Sem spadajo CDS, misc_feature itd.</p>
<p>Lahko izpišemo tudi nekatere druge karakteristike objekta (podrobneje o tem <a class="reference external" href="https://biopython.org/wiki/SeqRecord">tukaj</a>), na primer ID in opis:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>NM_005985.4
Homo sapiens snail family transcriptional repressor 1 (SNAI1), mRNA
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="shranjevanje-zapisa-v-datoteko">
<h3>Shranjevanje zapisa v datoteko<a class="headerlink" href="#shranjevanje-zapisa-v-datoteko" title="Permalink to this headline">¶</a></h3>
<p>Zapis lahko tudi shranimo v lokalno datoteko v formatu Genbank ali FASTA. Pozor, da spodnja koda deluje morate imeti v mapi, kjer se nahaja trenutno ta datoteka <code class="docutils literal notranslate"><span class="pre">ipynb</span></code>, ustvarjeno mapo <code class="docutils literal notranslate"><span class="pre">izhod</span></code>!</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">seq_record</span><span class="p">,</span> <span class="s1">&#39;izhod/genbank-sample_output.gb&#39;</span><span class="p">,</span> <span class="s1">&#39;gb&#39;</span><span class="p">)</span>   <span class="c1"># GenBank Flat File</span>
<span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">seq_record</span><span class="p">,</span> <span class="s1">&#39;izhod/genbank-sample_output.fasta&#39;</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)</span>   <span class="c1"># FASTA</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>1
</pre></div>
</div>
</div>
</div>
<p>Če bi iskanje ponavljali z različnimi kodami za dostop (pa tudi sicer) je smiselno, da kot ime datoteke uporabimo kar kodo zapisa v zbirko GenBank. To lahko naredimo tako:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">seq_record</span><span class="p">,</span> <span class="s1">&#39;izhod/genbank-</span><span class="si">%s</span><span class="s1">.gb&#39;</span> <span class="o">%</span> <span class="n">seq_record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s1">&#39;gb&#39;</span><span class="p">)</span>
<span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">seq_record</span><span class="p">,</span> <span class="s1">&#39;izhod/genbank-</span><span class="si">%s</span><span class="s1">.fasta&#39;</span> <span class="o">%</span> <span class="n">seq_record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>1
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="ekstrakcija-dela-zapisa-in-prevajanje">
<h3>Ekstrakcija dela zapisa in prevajanje<a class="headerlink" href="#ekstrakcija-dela-zapisa-in-prevajanje" title="Permalink to this headline">¶</a></h3>
<p>Zapis preberemo iz datoteke s <code class="docutils literal notranslate"><span class="pre">SeqIO.parse</span></code>, ekstrahiramo pa lahko določeno značilnost, na primer kodirajoče zaporedje (CDS) - seveda to velja na primer pri mRNA, značilnosti pa so prisotne samo v zapisu GenNank, pri FASTA formatu pa seveda ne. CDS lahko tudi prevedemo z uporabo določene kodne tabele (navodila na BioPython: <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc26">Translation Tables</a>) in dobimo nov objekt, iz katerega lahko izpišemo samo zaporedje (kako bo izpisan kodon STOP lahko tudi nastavimo).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;izhod/genbank-sample_output.gb&#39;</span><span class="p">,</span> <span class="s1">&#39;gb&#39;</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">rec</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">rec</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;CDS&#39;</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">location</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;protein_id&#39;</span><span class="p">])</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">rec</span><span class="p">)</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">rec</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="s1">&#39;Standard&#39;</span><span class="p">,</span> <span class="n">stop_symbol</span><span class="o">=</span><span class="s1">&#39;*&#39;</span><span class="p">))</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">rec</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="s1">&#39;Standard&#39;</span><span class="p">,</span> <span class="n">stop_symbol</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[80:875](+)
[&#39;NP_005976.2&#39;]
ATGCCGCGCTCTTTCCTCGTCAGGAAGCCCTCCGACCCCAATCGGAAGCCTAACTACAGCGAGCTGCAGGACTCTAATCCAGAGTTTACCTTCCAGCAGCCCTACGACCAGGCCCACCTGCTGGCAGCCATCCCACCTCCGGAGATCCTCAACCCCACCGCCTCGCTGCCAATGCTCATCTGGGACTCTGTCCTGGCGCCCCAAGCCCAGCCAATTGCCTGGGCCTCCCTTCGGCTCCAGGAGAGTCCCAGGGTGGCAGAGCTGACCTCCCTGTCAGATGAGGACAGTGGGAAAGGCTCCCAGCCCCCCAGCCCACCCTCACCGGCTCCTTCGTCCTTCTCCTCTACTTCAGTCTCTTCCTTGGAGGCCGAGGCCTATGCTGCCTTCCCAGGCTTGGGCCAAGTGCCCAAGCAGCTGGCCCAGCTCTCTGAGGCCAAGGATCTCCAGGCTCGAAAGGCCTTCAACTGCAAATACTGCAACAAGGAATACCTCAGCCTGGGTGCCCTCAAGATGCACATCCGAAGCCACACGCTGCCCTGCGTCTGCGGAACCTGCGGGAAGGCCTTCTCTAGGCCCTGGCTGCTACAAGGCCATGTCCGGACCCACACTGGCGAGAAGCCCTTCTCCTGTCCCCACTGCAGCCGTGCCTTCGCTGACCGCTCCAACCTGCGGGCCCACCTCCAGACCCACTCAGATGTCAAGAAGTACCAGTGCCAGGCGTGTGCTCGGACCTTCTCCCGAATGTCCCTGCTCCACAAGCACCAAGAGTCCGGCTGCTCAGGATGTCCCCGCTGA
ID: &lt;unknown id&gt;
Name: &lt;unknown name&gt;
Description: &lt;unknown description&gt;
Number of features: 0
/molecule_type=protein
Seq(&#39;MPRSFLVRKPSDPNRKPNYSELQDSNPEFTFQQPYDQAHLLAAIPPPEILNPTA...PR*&#39;)
MPRSFLVRKPSDPNRKPNYSELQDSNPEFTFQQPYDQAHLLAAIPPPEILNPTASLPMLIWDSVLAPQAQPIAWASLRLQESPRVAELTSLSDEDSGKGSQPPSPPSPAPSSFSSTSVSSLEAEAYAAFPGLGQVPKQLAQLSEAKDLQARKAFNCKYCNKEYLSLGALKMHIRSHTLPCVCGTCGKAFSRPWLLQGHVRTHTGEKPFSCPHCSRAFADRSNLRAHLQTHSDVKKYQCQACARTFSRMSLLHKHQESGCSGCPR
</pre></div>
</div>
</div>
</div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="naloga">
<h2>Naloga<a class="headerlink" href="#naloga" title="Permalink to this headline">¶</a></h2>
<div class="section" id="ozadje">
<h3>Ozadje<a class="headerlink" href="#ozadje" title="Permalink to this headline">¶</a></h3>
<p>Znano je , da se sestava kodirajoče in nekodirajoče regije nekoliko razlikuje, kar je opisano tudi na <a class="reference external" href="https://en.wikipedia.org/wiki/Coding_region">Wikipediji</a>: “<em>The coding region is thought to contain a higher GC-content than non-coding regions. There is further research that discovered that the longer the coding strand, the higher the GC-content. Short coding strands are comparatively still GC-poor, similar to the low GC-content of the base composition translational stop codons like TAG, TAA, and TGA.</em>”</p>
<p>Nekajna to temo lahko preberete tudi v članku <a class="reference external" href="https://dx.doi.org/10.1186/1471-2148-8-99">Both selective and neutral processes drive GC content evolution in the human genome</a>.</p>
</div>
<div class="section" id="id1">
<h3>Naloga<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>Vaša naloga je, da pripravite program, ki bo za dano kodo zapisa izračunal <strong>delež GC (v %) v CDS in pa v zaporedju, ki ni del CDS (torej v 5’- in 3’-neprevedeni regiji)</strong>.</p>
<p>Da bo analiza smiselna, je smiselno uporabiti genomsko zaporedje. Lahko za primer vzamete virusni genom SARS-CoV-2, en tak zapis je <a class="reference external" href="https://www.ncbi.nlm.nih.gov/nuccore/MT066156">GenBank MT066156</a> ali pa kratek genom kakšne bakterije (pobrskajte po GenBank in omejite rezultate na genomsko DNA dolžine med <span class="math notranslate nohighlight">\(10^6\)</span> in <span class="math notranslate nohighlight">\(1.5 \times 10^6\)</span> bp, da ne bo analiza trajala predolgo). Podite pozorni - genomi vsebujejo več kodirajočih regij, ki jih morate za to analizo združiti v eno zaporedje.</p>
</div>
<div class="section" id="namigi">
<h3>Namigi<a class="headerlink" href="#namigi" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Za izračun deležev GC si oglejte prvo (analiza kratkega nukleotidnega zaporedja), tam je tudi ilustrirano, kako dobite dolžino zaporedja. Izpis programa naj bo takšen, da bo jasno, kaj je kar - podobno, kot ste to naredili pri prvi vaji za izračun temperature tališča oligonukleotida.</p></li>
<li><p>Morda se naloga sliši kompleksna, a v resnici ni - s stavkom <code class="docutils literal notranslate"><span class="pre">if</span></code> lahko prečešete vhodno zaporedje za posamezne regije CDS in za vsako od njih izračunate število G+C, ki ga prištejete (uporabite <code class="docutils literal notranslate"><span class="pre">+=</span></code>) k skupnemu številu G+C.</p></li>
<li><p>Število G+C v nekodirajočih regijah lahko enostavno dobite tako, da izračunate število G+C v celotnem zaporedju, od katerega odštejete število G+C v vseh kodirajočih regijah.</p></li>
</ul>
<p>Predlagani koraki programa:</p>
<ol class="simple">
<li><p>Preberite zapis v formatu GenBank (<code class="docutils literal notranslate"><span class="pre">gb</span></code>).</p></li>
<li><p>Izračunajte dolžino celotnega zaporedja: <code class="docutils literal notranslate"><span class="pre">len</span></code> (primer pri <a class="reference internal" href="nt-oligo_analiza.html"><span class="doc std std-doc">V01</span></a>).</p></li>
<li><p>Preštejte G-je in C-je v celotnem zaporedju: <code class="docutils literal notranslate"><span class="pre">.count('C')</span></code> in <code class="docutils literal notranslate"><span class="pre">.count('G')</span></code> (primer pri <a class="reference internal" href="nt-oligo_analiza.html"><span class="doc std std-doc">V01</span></a>).</p></li>
<li><p>Definirajte dve spremenljivki (nastavite na <code class="docutils literal notranslate"><span class="pre">0</span></code>):</p>
<ul class="simple">
<li><p>za število G+C v vseh CDS (npr. <code class="docutils literal notranslate"><span class="pre">gc_cds</span></code>)</p></li>
<li><p>za skupno dolžino vseh CDS (npr. <code class="docutils literal notranslate"><span class="pre">len_cds</span></code>)</p></li>
</ul>
</li>
<li><p>Z zanko <code class="docutils literal notranslate"><span class="pre">for</span></code> (zgled zgoraj) prečešite vse CDS, pri tem pa vsakič število G+C v analiziranem CDS dodajte k vrednosti <code class="docutils literal notranslate"><span class="pre">gc_cds</span></code>, dolžino analiziranega CDS pa dodate k vrednosti <code class="docutils literal notranslate"><span class="pre">len_cds</span></code> (za dodajanje uporabite <code class="docutils literal notranslate"><span class="pre">+=</span></code>).</p></li>
<li><p>Število G+C v nekodirajočih regijah dobite tako, da od števila G+C v celotnem zaporedju odštejete <code class="docutils literal notranslate"><span class="pre">gc_cds</span></code> (to naredite šele, ko prečešete vse CDS).</p></li>
<li><p>Dolžino nekodirajočih regij dobite tako, da od celotne dolžine zaporedja odštejete vsoto dolžin vseh CDS (<code class="docutils literal notranslate"><span class="pre">len_cds</span></code>).</p></li>
<li><p>Na osnovi <code class="docutils literal notranslate"><span class="pre">gc_cds</span></code> in <code class="docutils literal notranslate"><span class="pre">len_cds</span></code> izračunajte delež GC v CDS.</p></li>
<li><p>Na osnovi G+C v nekodirajočih regijah in njihove skupne dolžine (izračunano zgoraj) izračunajte še delež GC v nekodirajočih regijah.</p></li>
</ol>
<p><em>Priporočam, da kode v zgledu ne spreminjate temveč si jo skopirate nekje tukaj spodaj ali v nov zvezek, tako da ohranite delujočo kodo za morebitni troubleshooting. Prav tako spremenite imena izhodnih datotek, ki jih program kreira.</em></p>
</div>
<div class="section" id="v-razmislek">
<h3>V razmislek<a class="headerlink" href="#v-razmislek" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Kaj pa, če analizirano zaporedje vsebuje prekrivajoče se CDS? Je vaš algoritem na to prilagojen? Če ne, kako bi ga prilagodili?</p></li>
<li><p>Pri analizi smo izračunali odstotek GC parov kar po vseh CDS hkrati (povprečenje), s tem pa smo nekaj informacij izgubili, npr. o razponu deleža GC ter o porazdelitvi deleža GC glede na dolžino CDS. Razmislite, kako bi analizo izvedli, da bi prišli do teh informacij in kako bi jih vizualizirali.</p></li>
</ul>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./vaje"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
            



<div class='prev-next-bottom'>
    
    <div id="prev">
        <a class="left-prev" href="genbank_web.html" title="previous page">
            <i class="prevnext-label fas fa-angle-left"></i>
            <div class="prevnext-info">
                <p class="prevnext-label">previous</p>
                <p class="prevnext-title">VAJA: Dostop do GenBank v brskalniku</p>
            </div>
        </a>
    </div>
     <div id="next">
        <a class="right-next" href="../osnove/ak.html" title="next page">
            <div class="prevnext-info">
                <p class="prevnext-label">next</p>
                <p class="prevnext-title">Aminokislinska zaporedja</p>
            </div>
            <i class="prevnext-label fas fa-angle-right"></i>
        </a>
     </div>

</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By Miha Pavšič<br/>
        
            &copy; Copyright 2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>